### PR TITLE
Add smithing script

### LINKED ISSUE #4: Smithing Script
Hey, this project is very intriguing. I've always wondered if something like this existed. I'd like to try my hand at making a smithing script that uses the Varrock west anvil. Are you accepting script submissions?

The script would start at Varrock west bank, withdraw the required items, walk to the anvil, smith the items based on the config, and walk back to the bank.

### PR DESCRIPTION
**Pull request by a-rokay**
This PR adds a smithing script that uses the Varrock west anvil and bank.

To accomplish this, I:
* Added a `smither()` function in `main.py`
* Added a Smithing class in `skills.py`
* Updated the example config to include smithing
* Added all the necessary needles for basic iron smithing in:
   * `needles/side-stones/inventory`
   * `needles/items`
   * `needles/game-screen/varrock`
   * `needles/game-screen/bank`
* Added a  Varrock west haystack in `/haystacks/`
* Added information to docs that explain the requirements

### Requirements
* Start next to southern most booth at Varrock west bank.
* The user's bank PIN must have already been entered.
* Must have 2 or more hammers banked.
* Bars and hammers must be visible without scrolling.

bank-setup

Fixes #4

### CODE REVIEW

File: ocvbot/main.py [hunk 1]
```diff
@@ -269,6 +269,58 @@ def test():
     banking.deposit_inventory()
 
 
+def smither(bar: str, smith: str, bars_per_item: int):
+    haystack_map = "./haystacks/varrock-west-bank.png"
+
+    # These are used to figure out when we're done smithing.
+    # Smithing items that take multiple bars may lead to 1 or 2 bars remaining.
+    inv_3_remaining = "./needles/side-stones/inventory/iron-bars-3.png"
```
[the remark this answers is not in the document]
Comment (takelley1) [reply ocvbot/main.py:277]: > What if the script sleeps a random amount of time with the lower limit being the normal amount of time it takes to smith a full inventory. If there's a level up message at the end of the sleep, we re-smith. That'd be more robust, take care of partial inventories, and possibly seem more human-like. What do you think?

That's definitely an option, but I'll leave it up to you since you've had the opportunity to test it more than I have at this point. The correct way to do this would definitely be to write a function that counts items in the inventory.

In general, "dumb" random sleeps aren't a good idea since they can't adapt to different network conditions. If you make the sleep long in order to account for slower connections, then on faster connections, the player would finish smithing and then just do nothing for a short while (not a big deal, but still). Alternately, if you make the sleep short in order to account for faster network connections, then on slower connections the script might try to return to the bank before the player is actually done smithing. This doesn't seem like much of a problem however, and it probably wouldn't raise any suspicion since other players might think you're simply smithing items that require more bars, so you would spend less time at the anvil.

I'm not certain how much network latency affects smithing speed, however -- it may not be affected by it at all. You could try calculating the amount of time it takes to smith an inventory and see how consistent it is. If you're on a laptop over Wi-fi, position the device at varying distances from your router to change the network latency. The wiki says that smithing a single item always takes 3 seconds (https://oldschool.runescape.wiki/w/Smithing#Smithing), but I would still verify this to be sure. If the smithing time is super consistent, then a dumb random wait would be fine. In the rare case of partial inventories, the player would simply do nothing at the anvil after smithing for a short while, which is fine. Obviously depending on the number of bars required to smith the desired item, you would wait for different periods of time.

[the remark this answers is not in the document]
Comment (a-rokay) [reply ocvbot/main.py:277]: Good point on the latency. Considering that, I believe the method I'm using now to be the best, until we come up with function to count items. I can add a check after withdrawing to make sure we have a full inventory. If we don't, stop the script. This will eliminate partial inventories.

[the remark this answers is not in the document]
Comment (takelley1) [reply ocvbot/main.py:277]: That should be fine. Sounds good!

**Comment by takelley1** [comment]
I finally had the opportunity to do a little testing, just have a few minor things:
- The `anvil.png` needle is a little small. I'd like it to cover the entire clickable area of the anvil, if possible.
- I tried smithing iron scimitars, and the script ended up clicking on iron swords, so the top border of the `iron-scimitar-bank.png` needle needs to be cropped by a few pixels. Make sure it covers the clickable area and no more.
- In the `config.yaml.example` file, list the number of bars required to smith each item.

**Comment by takelley1** [comment]
Another thing found after testing:
If you run out of iron bars, the script will attempt to withdrawal the empty item placeholder. To fix this, you'll have to add a `conf` parameter to `banking.withdrawal_item` and set it to something high, like `0.999`. Additionally, you should check if the `banking.withdrawal_item` function returns `False`. Otherwise, the script could fail to withdrawal the item and continue on to the anvil, attempting to smith with an empty inventory.

**Comment by a-rokay** [comment]
Working on 

> Another thing found after testing: If you run out of iron bars, the script will attempt to withdrawal the empty item placeholder. To fix this, you'll have to add a `conf` parameter to `banking.withdrawal_item` and set it to something high, like `0.999`. Additionally, you should check if the `banking.withdrawal_item` function returns `False`. Otherwise, the script could fail to withdrawal the item and continue on to the anvil, attempting to smith with an empty inventory.

Working on this now

**Comment by a-rokay** [comment]
I've fixed your concerns. For the placeholder bars, I added a check before withdrawing with a conf of 0.9999 (which seems to be perfect). Additionally, the script will retry to withdraw if `banking.withdrawal_item` returns false.

File: ocvbot/main.py [hunk 2]
```diff
@@ -269,6 +269,90 @@ def test():
...
+            needle=bar,
+            conf=0.9999
+        ).find_needle()
+
+        # Stop script if we don't
+        if have_bars is False:
+            log.info("Out of bars, stopping script.")
+            break
+
+        while withdrew_hammer is False:
```
Comment (takelley1) [inline ocvbot/main.py:334]: A `while` loop will continue forever to try withdrawing the hammer, even if it doesn't work. The `banking.withdrawal_item()` function will already try 5 times to withdrawal the item. It's better to do something like this:
```python
withdrew_hammer = banking.withdrawal_item(item_bank=hammer_bank, item_inv=hammer_inv, quantity="1")
if withdrew_hammer is False:
  log.error("Unable to withdrawal hammer!")
  break
```

**Review by takelley1 — CHANGES_REQUESTED** [review]

File: ocvbot/main.py (hunk 2 above)
Comment (a-rokay) [reply ocvbot/main.py]: My mistake, I missed that the withdraw function already executes multiple attempts. I've used your suggestion.

**Comment by takelley1** [comment]
I look forward to more contributions from you!

**Comment by a-rokay** [comment]
Thanks! Your comments made the experience very intuitive. I'll be back with some contributions for sure!
